File: src/sc_neurocore/accel/go/services/chialvo_map.go
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial license available
// © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
// © Code 2020–2026 Miroslav Šotek. All rights reserved.
// ORCID: 0009-0009-3560-0851
// Contact: www.anulum.li | protoscience@anulum.li
// SC-NeuroCore — Checked Go service for the Chialvo map

package services

import (
	"errors"
	"math"
)

// ChialvoMapNeuronState holds the two state variables and map parameters.
type ChialvoMapNeuronState struct {
	X          float64
	Y          float64
	A          float64
	B          float64
	C          float64
	K          float64
	XThreshold float64
}

// NewChialvoMapNeuron returns the source-paper parameter set used by Python.
func NewChialvoMapNeuron() *ChialvoMapNeuronState {
	return &ChialvoMapNeuronState{
		X:          0.0,
		Y:          0.0,
		A:          0.89,
		B:          0.6,
		C:          0.28,
		K:          0.04,
		XThreshold: 1.0,
	}
}

func finiteChialvoMap(value float64) bool {
	return !math.IsNaN(value) && !math.IsInf(value, 0)
}

func safeExpChialvoMap(value float64) float64 {
	return math.Exp(math.Min(500.0, math.Max(-500.0, value)))
}

// ValidateChialvoMap checks that every state and parameter field is finite.
func ValidateChialvoMap(state *ChialvoMapNeuronState) bool {
	if state == nil {
		return false
	}
	return finiteChialvoMap(state.X) &&
		finiteChialvoMap(state.Y) &&
		finiteChialvoMap(state.A) &&
		finiteChialvoMap(state.B) &&
		finiteChialvoMap(state.C) &&
		finiteChialvoMap(state.K) &&
		finiteChialvoMap(state.XThreshold)
}

// Step advances one simultaneous map iteration under an additive perturbation.
func (state *ChialvoMapNeuronState) Step(current float64) (int, error) {
	if !ValidateChialvoMap(state) {
		return 0, errors.New("invalid Chialvo map runtime state")
	}
	if !finiteChialvoMap(current) {
		return 0, errors.New("invalid Chialvo map current")
	}

	xPrevious := state.X
	xSquared := state.X * state.X
	exponential := safeExpChialvoMap(state.Y - state.X)
	xNext := xSquared*exponential + state.K + current
	yNext := state.A*state.Y - state.B*state.X + state.C
	if !finiteChialvoMap(xNext) || !finiteChialvoMap(yNext) {
		return 0, errors.New("invalid Chialvo map candidate state")
	}
	state.X = xNext
	state.Y = yNext
	if xPrevious < state.XThreshold && state.X >= state.XThreshold {
		return 1, nil
	}
	return 0, nil
}

// SimulateComplete runs a failure-atomic batch and returns both state traces.
func (state *ChialvoMapNeuronState) SimulateComplete(
	nSteps int,
	current float64,
) ([]float64, []float64, int64, error) {
	if nSteps < 0 {
		return nil, nil, 0, errors.New("nSteps must be non-negative")
	}
	if !ValidateChialvoMap(state) || !finiteChialvoMap(current) {
		return nil, nil, 0, errors.New("invalid Chialvo map batch input")
	}
	candidate := *state
	xTrace := make([]float64, nSteps)
	yTrace := make([]float64, nSteps)
	var spikes int64
	for index := range xTrace {
		spiked, err := candidate.Step(current)
		if err != nil {
			return nil, nil, 0, err
		}
		xTrace[index] = candidate.X
		yTrace[index] = candidate.Y
		spikes += int64(spiked)
	}
	state.X = candidate.X
	state.Y = candidate.Y
	return xTrace, yTrace, spikes, nil
}

// Simulate runs the complete checked batch and returns its fast-state trace.
func (state *ChialvoMapNeuronState) Simulate(
	nSteps int,
	current float64,
) ([]float64, int64, error) {
	xTrace, _, spikes, err := state.SimulateComplete(nSteps, current)
	return xTrace, spikes, err
}

// Reset restores state variables while preserving configured parameters.
func (state *ChialvoMapNeuronState) Reset() {
	state.X = 0.0
	state.Y = 0.0
}

// SimulateChialvoMapNeuron runs the default model for compatibility callers.
func SimulateChialvoMapNeuron(nSteps int, current float64) ([]float64, int64, error) {
	return NewChialvoMapNeuron().Simulate(nSteps, current)
}

File: src/sc_neurocore/accel/julia/neurons/chialvo_map.jl
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SC-NeuroCore — Checked Julia accelerator for the Chialvo map

module ChialvoMapAccel

export ChialvoMapNeuronState, reset!, simulate, simulate_trace, step!, validate

"""State and parameters for Chialvo's two-dimensional discrete map."""
mutable struct ChialvoMapNeuronState
    x::Float64
    y::Float64
    a::Float64
    b::Float64
    c::Float64
    k::Float64
    x_threshold::Float64
end

ChialvoMapNeuronState() = ChialvoMapNeuronState(0.0, 0.0, 0.89, 0.6, 0.28, 0.04, 1.0)

"""Return whether every state and parameter field is finite."""
function validate(state::ChialvoMapNeuronState)::Bool
    return isfinite(state.x) &&
           isfinite(state.y) &&
           isfinite(state.a) &&
           isfinite(state.b) &&
           isfinite(state.c) &&
           isfinite(state.k) &&
           isfinite(state.x_threshold)
end

safe_exp(value::Float64)::Float64 = exp(clamp(value, -500.0, 500.0))

"""Advance one simultaneous map iteration under an additive perturbation."""
function step!(state::ChialvoMapNeuronState, current::Float64 = 0.0)::Int64
    validate(state) || throw(ArgumentError("invalid Chialvo map runtime state"))
    isfinite(current) || throw(ArgumentError("invalid Chialvo map current"))

    x_previous = state.x
    x_squared = state.x * state.x
    exponential = safe_exp(state.y - state.x)
    x_next = x_squared * exponential + state.k + current
    y_next = state.a * state.y - state.b * state.x + state.c
    if !isfinite(x_next) || !isfinite(y_next)
        throw(OverflowError("invalid Chialvo map candidate state"))
    end

    state.x = x_next
    state.y = y_next
    return x_previous < state.x_threshold <= state.x ? Int64(1) : Int64(0)
end

"""Run checked iterations and return trace, event count, and final state."""
function simulate_trace(
    x0::Float64,
    y0::Float64,
    a::Float64,
    b::Float64,
    c::Float64,
    k::Float64,
    x_threshold::Float64,
    n_steps::Int,
    current::Float64,
)
    n_steps >= 0 || throw(ArgumentError("n_steps must be non-negative"))
    state = ChialvoMapNeuronState(x0, y0, a, b, c, k, x_threshold)
    validate(state) || throw(ArgumentError("invalid Chialvo map configuration"))
    trace = Vector{Float64}(undef, n_steps)
    spikes = Int64(0)
    for index in eachindex(trace)
        spikes += step!(state, current)
        trace[index] = state.x
    end
    return (; trace, spikes, xf = state.x, yf = state.y)
end

"""Run the default Chialvo parameter set for compatibility callers."""
function simulate(n_steps::Int = 1000; I_ext::Float64 = 0.0)
    result = simulate_trace(0.0, 0.0, 0.89, 0.6, 0.28, 0.04, 1.0, n_steps, I_ext)
    return result.trace, result.spikes
end

"""Restore state variables while preserving configured parameters."""
function reset!(state::ChialvoMapNeuronState)::Nothing
    state.x = 0.0
    state.y = 0.0
    return nothing
end

end # module ChialvoMapAccel
